cpp/src/strings/convert/convert_urls.cu

```diff
@@ -222,30 +286,84 @@ std::unique_ptr<column> url_decode(
+
+  // Count the number of URL escape sequences across all strings, ignoring string boundaries.
+  // This may count more sequences than actually are there since string boundaries are ignored.
+  size_type potential_esc_count =
+    thrust::count_if(rmm::exec_policy(stream),
+                     thrust::make_counting_iterator<size_type>(chars_start),
+                     thrust::make_counting_iterator<size_type>(chars_end),
+                     esc_detector);
+
+  if (potential_esc_count == 0) {
+    // nothing to replace, so just copy the input column
+    return std::make_unique<cudf::column>(strings.parent());
+  }
+
+  // create a vector of potential escape sequence positions
+  rmm::device_uvector<size_type> potential_esc_positions(potential_esc_count, stream);
+  auto d_potential_esc_positions = potential_esc_positions.data();
+  thrust::copy_if(rmm::exec_policy(stream),
+                  thrust::make_counting_iterator<size_t>(chars_start),
+                  thrust::make_counting_iterator<size_t>(chars_end),
+                  d_potential_esc_positions,
+                  esc_detector);
+
+  // Filter potential escape positions to actual positions by checking string boundaries.
+  // Skip counting the actual number to compute the optimal output buffer size since it is not
+  // expected to be significantly smaller in practice.
+  rmm::device_uvector<size_type> esc_positions(potential_esc_count, stream);
+  auto d_esc_positions = potential_esc_positions.data();
+  auto esc_pos_end     = thrust::copy_if(rmm::exec_policy(stream),
+                                     d_potential_esc_positions,
+                                     d_potential_esc_positions + potential_esc_count,
+                                     d_esc_positions,
+                                     url_decode_esc_position_filter{strings.size() + 1, d_offsets});
+  size_type esc_count  = esc_pos_end - d_esc_positions;
+
+  if (esc_count == 0) {
+    // nothing to replace, so just copy the input column
+    return std::make_unique<cudf::column>(strings.parent());
+  }
 
-  // copy null mask
-  rmm::device_buffer null_mask = cudf::detail::copy_bitmask(strings.parent(), stream, mr);
   // build offsets column
-  auto offsets_transformer_itr = thrust::make_transform_iterator(
-    thrust::make_counting_iterator<size_type>(0), url_decoder_fn{d_strings});
-  auto offsets_column = make_offsets_child_column(
-    offsets_transformer_itr, offsets_transformer_itr + strings_count, stream, mr);
-  auto d_offsets = offsets_column->view().data<int32_t>();
+  auto offsets_column = make_numeric_column(
+    data_type{type_id::INT32}, strings_count + 1, mask_state::UNALLOCATED, stream, mr);
+  auto offsets_view = offsets_column->mutable_view();
+  thrust::transform(rmm::exec_policy(stream),
+                    strings.offsets().begin<size_type>(),
+                    strings.offsets().end<size_type>(),
+                    offsets_view.begin<size_type>(),
```

Comment:
```suggestion
                    strings.offsets().begin<int32_t>(),
                    strings.offsets().end<int32_t>(),
                    offsets_view.begin<int32_t>(),
```
The offsets are created using `INT32` but referenced as `size_type`
This is potentially dangerous if `size_type` ever changes.